keps/sig-storage/2268-non-graceful-shutdown/README.md

```diff
@@ -0,0 +1,457 @@
+  1. new `quarantine` taint has NoSchedule effect, meaning it will not schedule on the shutdown node. For example, `quarantine=nodeshutdown:NoSchedule` or `quarantine=hardwarefailure:NoSchedule`. Although both `NoSchedule` and ``NoExecute` will not schedule new pods to the shutdown node, the difference is `NoExecute` will also evict pods unless pods tolerate this effect. So `NoSchedule` is chosen in this proposal.
+  1. We'll follow taint and toleration policy. If a pod is set to tolerate all taints and effects, that means user does NOT want to evict pods when node is not ready. So GC controller will filter out those pods and only forcefully delete pods that do not have a matching toleration.
+
+1. [Proposed change] Once pods are selected and forcefully deleted, the attachdetach reconciler should check the `quarantine` taint on the node. If the taint is present, the attachdetach reconciler will not wait for 6 minutes to do force detach. Instead it will force detach right away and allow `volumeAttachment` to be deleted.
+
+1. This would trigger the deletion of the `volumeAttachment` objects. For CSI drivers, this would allow `ControllerUnpublishVolume` to happen without `NodeUnpublishVolume` and/or `NodeUnstageVolume` being called first. Note that there is no additional code changes required for this step. This happens automatically after the `Proposed change` in the previous step to force detach right away.
+
+1. When the `external-attacher` detects the `volumeAttachment` object is being deleted, it calls CSI driver's `ControllerUnpublishVolume`.
+
+### Handle the Return of Shutdown Node
+
+If the nodes that were previously shutdown with pods forcefully deleted come back, we need to make sure they are cleaned up before pods are scheduled on them again.
+
+[Proposed Change] We require the user to manually remove the `quarantine` taint after the pods are moved to a new node since the user was the one who originally added the taint. This makes sure no new pods are scheduled on the node before the `quarantine` taint is removed. In the future, we can enhance the Pod GC Controller to automatically remove the `quarantine` taint, after verifying that no pods are attached to the node, and no volume attachments on the node.
+
+Note that there is no additional code change required for the return of shutdown node.
+
+[Existing Logic] The health check in Node lifecycle controller, part of kube-controller-manager, sets Node v1.NodeReady Condition to True if lease is renewed again. Node Status becomes Ready.
+
+[Existing Logic] Note that if the `quarantine` taint was not added to the shutdown node, the pods will be stuck in terminating status forever if the shutdown node remains down. If the shutdown node comes up, however, the original `volumeAttachment` will be deleted and the pods will be moved to a new node successfully.
```

Comment:
Who deletes the original VolumeAttachment in the existing logic?

Reply:
attach detach controller does.